rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+such as Credential Activation, proving that the TPM is legitimate and that you 
+are talking directly to it.
+
+## Join Process
+
+### Pre-join Configuration
+
+Before joining is possible, the user must configure the Auth Server with a join
+token which allows that TPM to join. The Join Token will contain the EKPub
+hashes and EKCert serials which are allowed to join.
+
+To assist in this process, a new command will be added `tbot tpm identify`:
+
+1. User runs `tbot tpm identify` on the host. This will query the EKPub and
+  EKCert from the TPM attached to the host and output these values for the user.
+2. User creates or modifies an existing `tpm` join token to include the EKPub
+  or EkCert serial of the TPM within the allow list. See the API Changes section
+  for the structure of this.
+
+### Joining
+
+The join process is roughly based on the TCG TPM 2.0 Keys for Device Identity
+and Attestation recommended "Identity Provisioning" flow. As per this document,
+the flow is designed to assure that:
+
+- The TPM is authentic (by validating the EKCert)
+- The specific TPM is authentic and is represented by the EKCert (by validating
+  that the EK in the TPM corresponds to the key in the EKCert)
+- The generated AK is authentic (resident in the same TPM as the EK)
+
+```mermaid
+sequenceDiagram
+  participant T as TPM
+  participant B as Bot
+  participant A as Auth Server
+  B->>T: Queries EK and EKCert
+  T-->>B: EK, EK handle, EKCert
+  B->>T: Request AK generation
+  T-->>B: AKPub, AK handle
+  B->>A: Starts RegisterUsingTPM RPC stream
+  B->>A: Sends EKPub, EKCert and AKPub
+  opt TPM CA configured for Join Token
+    A->>A: EKCert signature validated against CA
+  end
+  A->>A: Validates EKPub hash, EKCert serial against allow rules in Join Token 
+  A->>A: Generates Credential Activation challenge
+  A-->>B: Sends Credential Activation challenge
+  B->>T: ActivateCredential(EK Handle, AK Handle, Challenge)
+  T->>T: Solves challenge
+  T-->>B: Credential Activation solution
+  B-->>A: Credential Activation solution
+  A->>A: Validates submitted solution against the known solution
+  A->>B: Signed TLS and SSH Certificates
+```
+
+At the core of this process is the Credential Activation ceremony. This proves
+possession of the EK and AK to the Auth Server through decrypting an
+encrypted challenge. As long as the TPM is authentic (e.g the EKCert is 
+signed by a trusted manufacturer), this ceremony also proves that the EK and AK
+are resident in the same TPM and cannot be exported.
+
+### Renewal
+
+As with all delegated join methods, the renewal will simply repeat the join
+process.
+
+## API Changes
+
+### Join Token
+
+A new JoinMethod will be added - `tpm`.
+
+The existing `JoinToken` type will be extended to include additional fields for
+TPM joining:
+
+```protobuf
+// ProvisionTokenSpecV2TPM contains the TPM-specific part of the
+// ProvisionTokenSpecV2
+message ProvisionTokenSpecV2TPM {
```

Comment:
nit: Do we have to version the message name? Could this be ProvisionTokenSpecTPM?

Renaming is a breaking change, so it's likely we'll never actually have a "V3".

Reply:
😓  So far all of the JoinMethod specific parts of the `ProvisionTokenSpecV2` have included the `V2` - for the sake of consistency it feels like we ought to continue that.

Reply:
That's alright, thanks for humoring me.